In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

# Database Info

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

 * sqlite:///chinook.db
Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

### Relational Model

![title](db_relational_model.svg)

# Top 3 Genres sold in USA

In [14]:
%%sql
SELECT 
                il.invoice_id, 
                g.name AS genre
                FROM invoice_line il
                INNER JOIN track t ON il.track_id = t.track_id
                INNER JOIN genre g ON t.genre_id = g.genre_id
            WHERE 
                genre LIKE '%Hip%' 
                    OR genre LIKE '%Punk%' 
                      OR genre LIKE '%Pop%' 
                        OR genre LIKE '%Blues%';

 * sqlite:///chinook.db
Done.


[(2, 'Pop'),
 (2, 'Alternative & Punk'),
 (2, 'Alternative & Punk'),
 (2, 'Blues'),
 (7, 'Pop'),
 (7, 'Alternative & Punk'),
 (9, 'Blues'),
 (11, 'Alternative & Punk'),
 (11, 'Alternative & Punk'),
 (14, 'Alternative & Punk'),
 (17, 'Alternative & Punk'),
 (19, 'Alternative & Punk'),
 (20, 'Blues'),
 (20, 'Alternative & Punk'),
 (21, 'Alternative & Punk'),
 (25, 'Alternative & Punk'),
 (25, 'Alternative & Punk'),
 (28, 'Alternative & Punk'),
 (29, 'Alternative & Punk'),
 (30, 'Blues'),
 (30, 'Alternative & Punk'),
 (30, 'Alternative & Punk'),
 (33, 'Hip Hop/Rap'),
 (35, 'Pop'),
 (36, 'Alternative & Punk'),
 (37, 'Alternative & Punk'),
 (38, 'Pop'),
 (38, 'Alternative & Punk'),
 (40, 'Alternative & Punk'),
 (42, 'Blues'),
 (43, 'Blues'),
 (43, 'Alternative & Punk'),
 (45, 'Alternative & Punk'),
 (47, 'Alternative & Punk'),
 (51, 'Alternative & Punk'),
 (51, 'Alternative & Punk'),
 (55, 'Alternative & Punk'),
 (59, 'Blues'),
 (60, 'Alternative & Punk'),
 (63, 'Pop'),
 (65, 'Alternative & Punk'),
 (65, 'Alternative & Punk'),
 (66, 'Alternative & Punk'),
 (67, 'Alternative & Punk'),
 (67, 'Pop'),
 (68, 'Blues'),
 (68, 'Blues'),
 (78, 'Alternative & Punk'),
 (81, 'Blues'),
 (86, 'Alternative & Punk'),
 (87, 'Alternative & Punk'),
 (88, 'Alternative & Punk'),
 (91, 'Alternative & Punk'),
 (91, 'Alternative & Punk'),
 (91, 'Alternative & Punk'),
 (94, 'Alternative & Punk'),
 (95, 'Alternative & Punk'),
 (95, 'Alternative & Punk'),
 (95, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (99, 'Alternative & Punk'),
 (102, 'Pop'),
 (102, 'Alternative & Punk'),
 (106, 'Pop'),
 (106, 'Alternative & Punk'),
 (108, 'Alternative & Punk'),
 (109, 'Alternative & Punk'),
 (110, 'Alternative & Punk'),
 (111, 'Alternative & Punk'),
 (113, 'Alternative & Punk'),
 (117, 'Hip Hop/Rap'),
 (117, 'Alternative & Punk'),
 (117, 'Alternative & Punk'),
 (117, 'Hip Hop/Rap'),
 (120, 'Alternative & Punk'),
 (124, 'Alternative & Punk'),
 (124, 'Alternative & Punk'),
 (124, 'Alternative & Punk'),
 (125, 'Blues'),
 (127, 'Alternative & Punk'),
 (129, 'Alternative & Punk'),
 (131, 'Blues'),
 (134, 'Blues'),
 (136, 'Hip Hop/Rap'),
 (139, 'Alternative & Punk'),
 (140, 'Alternative & Punk'),
 (143, 'Blues'),
 (143, 'Alternative & Punk'),
 (143, 'Alternative & Punk'),
 (143, 'Alternative & Punk'),
 (144, 'Blues'),
 (145, 'Pop'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (146, 'Alternative & Punk'),
 (147, 'Alternative & Punk'),
 (148, 'Alternative & Punk'),
 (151, 'Pop'),
 (154, 'Hip Hop/Rap'),
 (155, 'Alternative & Punk'),
 (155, 'Alternative & Punk'),
 (158, 'Alternative & Punk'),
 (158, 'Alternative & Punk'),
 (160, 'Pop'),
 (160, 'Alternative & Punk'),
 (161, 'Alternative & Punk'),
 (161, 'Blues'),
 (162, 'Alternative & Punk'),
 (162, 'Blues'),
 (162, 'Hip Hop/Rap'),
 (163, 'Alternative & Punk'),
 (163, 'Alternative & Punk'),
 (164, 'Blues'),
 (165, 'Hip Hop/Rap'),
 (165, 'Alternative & Punk'),
 (166, 'Pop'),
 (166, 'Alternative & Punk'),
 (167, 'Alternative & Punk'),
 (167, 'Alternative & Punk'),
 (168, 'Pop'),
 (169, 'Alternative & Punk'),
 (169, 'Alternative & Punk'),
 (171, 'Alternative & Punk'),
 (171, 'Alternative & Punk'),
 (171, 'Alternative & Punk'),
 (171, 'Alternative & Punk'),
 (

In [28]:
%%sql
WITH 
    invoice_usa AS 
        (
            SELECT
                c.country,
                i.invoice_id,
                i.total    
                FROM customer c
                INNER JOIN invoice i on c.customer_id = i.customer_id
            WHERE c.country = 'USA'
        ),
    top_genres AS 
    (       
            SELECT 
                il.invoice_id, 
                g.name AS genre
                FROM invoice_line il
                INNER JOIN track t ON il.track_id = t.track_id
                INNER JOIN genre g ON t.genre_id = g.genre_id
            WHERE 
                genre LIKE '%Hip%' 
                    OR genre LIKE '%Punk%' 
                      OR genre LIKE '%Pop%' 
                        OR genre LIKE '%Blues%'
    )
    

SELECT 
    tg.genre AS Genre,
    ROUND(SUM(iu.total), 0) Tracks_Sold,
    ROUND((SUM(iu.total) / (SELECT SUM(total) FROM invoice_usa)), 2) AS Percentage,
    iu.country AS Country
    FROM invoice_usa iu
    INNER JOIN top_genres tg ON iu.invoice_id = tg.invoice_id
GROUP BY 1
ORDER BY 2 DESC
    

    

 * sqlite:///chinook.db
Done.


[('Alternative & Punk', 1235.0, 1.19, 'USA'),
 ('Blues', 453.0, 0.44, 'USA'),
 ('Hip Hop/Rap', 366.0, 0.35, 'USA'),
 ('Pop', 202.0, 0.19, 'USA')]

# Employee Sales Performance

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase, let us analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

In [4]:
%%sql
SELECT * 
    FROM employee
LIMIT 3;

 * sqlite:///chinook.db
Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com')]

In [36]:
%%sql 
WITH 
    sales_employees AS
        (
            SELECT 
                e.employee_id,
                e.first_name || " " || e.last_name AS employee_name,
                e.title,
                e.hire_date, 
                e.reports_to
            FROM employee e
            WHERE e.title = 'Sales Support Agent'
        ),
    sales_performance AS
        (
            SELECT 
                se.employee_id,
                COUNT(c.customer_id) AS customer_number,
                COUNT(DISTINCT(c.country)) AS number_countries,
                ROUND(SUM(i.total), 2) AS total_sales
                FROM customer c
                INNER JOIN invoice i ON i.customer_id = c.customer_id
                INNER JOIN sales_employees se ON se.employee_id = c.support_rep_id
            GROUP BY 1
            ORDER BY 3 DESC
        ),
    sales_managers AS
        (
            SELECT 
                e.employee_id,
                e.first_name || " " || e.last_name AS employee_name,
                e.title,
                e.hire_date, 
                e.reports_to
            FROM employee e
            WHERE e.title = 'Sales Manager'
        )
    
SELECT 
    se.employee_name,
    se.title,
    se.hire_date,
    sm.employee_name AS manager,
    sp.total_sales,
    ROUND((sp.total_sales / (SELECT SUM(total_sales) FROM sales_performance)),2) 'sales_%', 
    sp.customer_number,
    sp.number_countries
    FROM sales_performance sp
    INNER JOIN sales_employees se ON se.employee_id = sp.employee_id
    INNER JOIN sales_managers sm ON sm.employee_id = se.reports_to
ORDER BY 5 DESC, 6 DESC;

 * sqlite:///chinook.db
Done.


[('Jane Peacock', 'Sales Support Agent', '2017-04-01 00:00:00', 'Nancy Edwards', 1731.51, 0.37, 212, 10),
 ('Margaret Park', 'Sales Support Agent', '2017-05-03 00:00:00', 'Nancy Edwards', 1584.0, 0.34, 214, 12),
 ('Steve Johnson', 'Sales Support Agent', '2017-10-17 00:00:00', 'Nancy Edwards', 1393.92, 0.3, 188, 13)]

**Jane Peacock** has the maximum total sales but we should take into consideration that he is the oldest one on the department. **Margaret Park** has an extra number of **customer & country** also **Steve Johnson the newest** one have customers more countries.
Another observation is that the **Top 3 Sales support agents** are in the same team or at least report to **Nancy Edwards** whose job to keep them on track and in a healthy competitive environment may be essential.

# Sales Performance By Country

We are going to make sales analysis for customers from each different country (countries with only one customer, will be grouped in "Other" category). The analysis will include the following points:
<ul>    
    <li>Total number of customers.</li>   
    <li>Total value of sales.</li>    
    <li>Average value of sales per customer.</li>    
    <li>Average order value.</li>
</ul>

In [52]:
%%sql

WITH 
    customers_info AS
        (
            SELECT
                c.customer_id,                
                c.country,
                i.total,
                i.invoice_id
                FROM customer c
                INNER JOIN invoice i ON i.customer_id = c.customer_id
            GROUP BY 1
            ORDER BY 1 
        ),
    country_sales AS
        (
            SELECT
                COUNT(ci.customer_id) AS number_of_customers,
                COUNT(DISTINCT(ci.invoice_id)) AS number_of_orders,
                ROUND(SUM(ci.total), 2) AS total_sales,
                CASE 
                    WHEN COUNT(DISTINCT(ci.customer_id)) > 1
                        THEN ci.country
                    ELSE 'Other'
                END AS country
                FROM customers_info ci
            GROUP BY ci.country
            ORDER BY 1 DESC
            
        ),
    country_sales_summary AS
        (
            SELECT 
                cs.country,
                SUM(cs.number_of_customers) AS total_customers,
                SUM(cs.number_of_orders) AS total_orders,
                SUM(cs.total_sales) AS total_sales,
                CASE
                    WHEN cs.country = 'Other' THEN 1
                    ELSE 0
                END AS sort
            FROM country_sales cs
            GROUP BY 1
            ORDER BY sort ASC, total_sales DESC
        )


SELECT 
    country,
    total_customers,
    total_orders,
    ROUND(total_sales, 2) AS total_sales,
    ROUND((total_sales / total_customers), 2) AS average_sales_per_customer,
    ROUND((total_sales / total_orders), 2) AS average_order_value
FROM country_sales_summary;



 * sqlite:///chinook.db
Done.


[('USA', 13, 13, 117.81, 9.06, 9.06),
 ('Canada', 8, 8, 68.31, 8.54, 8.54),
 ('Germany', 4, 4, 42.57, 10.64, 10.64),
 ('Brazil', 5, 5, 36.63, 7.33, 7.33),
 ('France', 5, 5, 32.67, 6.53, 6.53),
 ('United Kingdom', 3, 3, 30.69, 10.23, 10.23),
 ('Czech Republic', 2, 2, 25.74, 12.87, 12.87),
 ('India', 2, 2, 13.86, 6.93, 6.93),
 ('Portugal', 2, 2, 5.94, 2.97, 2.97),
 ('Other', 15, 15, 115.83, 7.72, 7.72)]

<ul>   
    <li>USA have the maximum number of sales with total sales 1040 taking about 22% of our sales , but we have about 13 customers so the sales average per customer is about 80 with sales average per order 7.94.</li>
    <li>Czech Republic have the Maximum sales average per customer 136.62 and by order 9.11 although it takes only 6% of the total sales.</li>
<li>United Kingdom, Portugal & India show an opportunity to increase the number of customers, sales there as sales average per customer & average order also is better than USA with few exceptions.</li>
</ul>

# Album vs Individual Tracks

In [29]:
%%sql
WITH 
    invoice_info AS
        (
            SELECT 
                t.album_id, 
                il.track_id, 
                il.invoice_id
                FROM track t
                INNER JOIN invoice_line il ON il.track_id = t.track_id
            GROUP BY 3
        )

SELECT 
    whole_album,
    count(distinct invoice_id) number_invoices, 
    cast(count(distinct invoice_id) as Float)/
                   (SELECT count(distinct invoice_id) FROM invoice) perc_inv
FROM 
    (
    SELECT inv.*, 
           CASE
            WHEN 
                (
                      SELECT t2.track_id /* whole album */
                          FROM track t2
                          WHERE t2.album_id = inv.album_id
                  
                      EXCEPT
                  
                      SELECT il2.track_id /* only purchased tracks */
                          FROM invoice_line il2
                          WHERE il2.invoice_id = inv.invoice_id
                 ) IS NULL
                    
                AND
        
                (
                      SELECT il2.track_id
                          FROM invoice_line il2
                          WHERE il2.invoice_id = inv.invoice_id
                      
                      EXCEPT
                    
                      SELECT t2.track_id
                          FROM track t2
                          WHERE t2.album_id = inv.album_id                
                ) IS NULL
                THEN 'Yes'
                ELSE 'No'
              END AS whole_album
    FROM invoice_info AS inv
    )
GROUP BY whole_album
    

 * sqlite:///chinook.db
Done.


[('No', 500, 0.8143322475570033), ('Yes', 114, 0.18566775244299674)]

# Top Artist in playlists

In [16]:
%%sql

WITH 
    play_lists_tracks AS 
        (
            SELECT
                pl.playlist_id,
                pl.track_id,
                t.album_id,
                al.artist_id
                FROM playlist_track pl
                INNER JOIN track t ON t.track_id = pl.track_id
                INNER JOIN album al ON al.album_id = t.album_id
            ORDER BY al.artist_id                                                
        )

        
SELECT     
    ar.artist_id AS ARTIST_ID,    
    ar.name AS ARTIST_NAME,
    COUNT(plt.track_id) AS PLAYLIST_APPEARANCE
    FROM artist ar
    INNER JOIN play_lists_tracks plt ON plt.artist_id = ar.artist_id    
GROUP BY ar.artist_id
ORDER BY 3 DESC
LIMIT 3;

 * sqlite:///chinook.db
Done.


[(90, 'Iron Maiden', 516), (150, 'U2', 333), (50, 'Metallica', 296)]

The artist that most appeared in playlistst is **Iron Maiden** followed by **U2 and Metallica**.

# Tracks Purchased VS Not Purchased 

In [20]:
%%sql 

WITH
    purchase_info AS
        (
            SELECT
                t.track_id,
                CASE
                    WHEN 
                        (
                          t.track_id IN (SELECT il.track_id FROM invoice_line il)      
                        ) 
                    THEN "Yes"
                    ELSE "No"                    
                END AS purchased                                    
                FROM track t
        )
SELECT 
    pi.purchased,
    COUNT(pi.track_id) AS number_of_tracks,
    ROUND((CAST(COUNT(pi.track_id) AS FLOAT) * 100) / (SELECT COUNT(*) FROM track t), 2) AS percentage
    FROM purchase_info pi
GROUP BY 1

 * sqlite:///chinook.db
Done.


[('No', 1697, 48.44), ('Yes', 1806, 51.56)]

Only the **51.56** of the tracks have been purchased.

# Is the range of tracks in the store reflective of their sales popularity?

For each genre we will extract the total percentage of tracks in store and the percentage sold.

In [65]:
%%sql
WITH 
    number_of_tracks_in_store AS 
        (
            SELECT COUNT(t.track_id) AS total
            FROM track t
        ),      
    number_of_tracks_sold AS
        (
            SELECT COUNT(il.track_id) AS total
            FROM invoice_line il
        ),  
    store_tracks AS
        (
            SELECT 
                t.genre_id,
                g.name AS genre_name,
                COUNT(t.track_id) total_tracks_in_store,
                ROUND(CAST(COUNT(t.track_id)AS FLOAT)*100/(SELECT COUNT(t.track_id)FROM track t),2) AS percentage_of_tracks_in_store    
                FROM track t
                INNER JOIN genre g ON g.genre_id =t.genre_id
            GROUP BY 2             
        ),
    store_tracks_purchased AS
        (
            SELECT 
                t.genre_id,
                g.name AS genre_name,
                COUNT(il.track_id) total_tracks_sold,
                ROUND(CAST(COUNT(il.track_id)AS FLOAT)*100/(SELECT COUNT(il.track_id)FROM invoice_line il),2) AS percentage_of_tracks_sold
                FROM invoice_line il 
                INNER JOIN track t ON il.track_id =t.track_id
                INNER JOIN genre g ON g.genre_id = t.genre_id
            GROUP BY 2 
        )
    
        
SELECT 
    st.genre_name AS genre,
    st.percentage_of_tracks_in_store,
    stp.percentage_of_tracks_sold
    FROM store_tracks st
    INNER JOIN store_tracks_purchased stp ON stp.genre_id = st.genre_id
ORDER BY 3 DESC
        


 * sqlite:///chinook.db
Done.


[('Rock', 37.03, 55.39),
 ('Metal', 10.68, 13.01),
 ('Alternative & Punk', 9.48, 10.34),
 ('Latin', 16.53, 3.51),
 ('R&B/Soul', 1.74, 3.34),
 ('Blues', 2.31, 2.61),
 ('Jazz', 3.71, 2.54),
 ('Alternative', 1.14, 2.46),
 ('Easy Listening', 0.69, 1.56),
 ('Pop', 1.37, 1.32),
 ('Electronica/Dance', 0.86, 1.16),
 ('Classical', 2.11, 0.99),
 ('Reggae', 1.66, 0.74),
 ('Hip Hop/Rap', 1.0, 0.69),
 ('Heavy Metal', 0.8, 0.17),
 ('Soundtrack', 1.23, 0.11),
 ('TV Shows', 2.65, 0.04),
 ('Drama', 1.83, 0.02)]